sim-administration/uploader/outfile_to_hss_input_converter.go

```diff
@@ -0,0 +1,336 @@
+	return state.headerDescription["Batch No"]
+}
+
+func getProfileType(state ParserState) string {
+	return state.headerDescription["ProfileType"]
+}
+
+func getCustomer(state ParserState) string {
+	// TODO: Maker safe, so that it fails reliably if Customer is not in map.
+	//       also use constant, not magic string
+	return state.headerDescription["Customer"]
+}
+
+func parseOutputLine(s string) (string, string, string) {
+	parsedString := strings.Split(s, " ")
+	return parsedString[0], parsedString[1], parsedString[2]
+}
+
+func transitionMode(state *ParserState, targetState string) {
+	state.currentState = targetState
+}
+
+func modeFromSectionHeader(s string) string {
+	sectionName := s[1:len(s)]
+	if sectionName == "HEADER DESCRIPTION" {
+		return HEADER_DESCRIPTION
+	} else if sectionName == "INPUT VARIABLES" {
+		return INPUT_VARIABLES
+	} else if sectionName == "OUTPUT VARIABLES" {
+		return OUTPUT_VARIABLES
+	} else {
+		return UNKNOWN_HEADER
+	}
+}
+
+func isSectionHeader(s string) bool {
+	match, _ := regexp.MatchString("^\\*([A-Z0-9 ])+$", s)
+	return match
+}
+
+func isComment(s string) bool {
+	match, _ := regexp.MatchString("^\\*+$", s)
+	return match
+}
+
+/// XXX Add a main function that
+//   a) Reads the output file, then produces a HSS input file from that
+//   b) Later, integrate with the prime input generator, and add a
+//      database to keep track of the workflow.
+//
+//
+
+//
+// Set up command line parsing
+//
+func parseCommandLine() (string, string) {
+	inputFile := flag.String("input-file",
+		"not  a valid filename",
+		"path to .out file used as input file")
+
+	outputFile := flag.String("output-file-prefix",
+		"not  a valid filename",
+		"prefix to path to .csv file used as input file, filename will be autogenerated")
+
+	//
+	// Parse input according to spec above
+	//
+	flag.Parse()
+	return *inputFile, *outputFile
+}
+
+// fileExists checks if a file exists and is not a directory before we
+// try using it to prevent further errors.
+func fileExists(filename string) bool {
+	info, err := os.Stat(filename)
+	if os.IsNotExist(err) {
+		return false
+	}
+	return !info.IsDir()
+}
+
+func WriteHssCsvFile(filename string, entries []SimEntry) error {
+
+	if fileExists(filename) {
+		log.Fatal("Output file already exists. '", filename, "'.")
+	}
+
+	f, err := os.Create(filename)
+	if err != nil {
+		log.Fatal("Couldn't create hss csv file '", filename, "': ", err)
+	}
+
+	_, err = f.WriteString("ICCID, IMSI, KI\n")
+	if err != nil {
+		log.Fatal("Couldn't header to  hss csv file '", filename, "': ", err)
+	}
+
+	max := 0
+	for i, entry := range entries {
+		s := fmt.Sprintf("%s, %s, %s\n", entry.iccidWithChecksum, entry.imsi, entry.ki)
+		_, err = f.WriteString(s)
+		if err != nil {
+			log.Fatal("Couldn't write to  hss csv file '", filename, "': ", err)
+		}
+		max = i + 1
+	}
+	fmt.Println("Successfully written ", max, " sim card records.")
+	return f.Close()
+}
+
+///
+///   Main.
+///
+
+func main() {
```

Comment:
I've found it quite common among Go programmers to put `main` at the top of the file, and more generally to list functions as a breadth-first traversal of the call tree (in the absence of recursive calls).  The rationale being that, when you open a file, the first thing you want to see is the highest-level entry point, which is much more relevant than the implementation details of helper functions.

Reply:
ok